### PR TITLE
Add scoped topics, suback response topic before publish

### PR DESCRIPTION
**Pull request by jw-msft**
- Distinguishes between scoped topics (operations other than enrollment) and non-scoped topics (enrollment response will contain the scope suffix for later operations)
- Matches correlation id from the mqtt correlation-data to pair the subscripbe with its acknowledgement from the mqtt broker
- various fixes and tweaks
- deletion of some unused/redundant codec
- made state store in-memory only to simplify (will add persistence back later)

### CODE REVIEW

**Comment by Jerry-GeGarcia** [comment]
>  *  For 'ainfo_resp' messages, if the Correlation Data matches the 'en,the client should parse the message and update the agentinfo state.

what is 'en?

---
Refers to: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c:65 in 4720e51. [](commit_id = 4720e512a20cf4eb0d01d8f7f504bbe837449c27, deletion_comment = False)

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c
Comment (Jerry-GeGarcia) [inline src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c:159]: trace worthy?

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_enrollment_management.c
Comment (Jerry-GeGarcia) [inline src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_enrollment_management.c:129]: >OnPublish_enr_resp [](http://example.com/codeflow?start=5&length=18)

this is almost a copy of OnPublish_ainfo_resp; some refactoring in order?

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/agentinfo_request_operation.c
Comment (Jerry-GeGarcia) [inline src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/agentinfo_request_operation.c:222]: > operationTimeoutSecs set correctly to non-zero [](http://example.com/codeflow?start=69&length=47)

server side?

File: src/extensions/agent_modules/shared/common/inc/aduc/adu_module_state.h
Comment (Jerry-GeGarcia) [inline src/extensions/agent_modules/shared/common/inc/aduc/adu_module_state.h:19]: >we're [](http://example.com/codeflow?start=47&length=5)

we only

File: src/extensions/agent_modules/shared/common/inc/aduc/adu_module_state.h
Comment (Jerry-GeGarcia) [inline src/extensions/agent_modules/shared/common/inc/aduc/adu_module_state.h:18]: >subscribe [](http://example.com/codeflow?start=58&length=9)

subscribed

**Comment by Jerry-GeGarcia** [comment]
>  * @brief Implements the Device Update agent state store functionality.

I could use another pair of eyes on this

---
Refers to: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:3 in 4720e51. [](commit_id = 4720e512a20cf4eb0d01d8f7f504bbe837449c27, deletion_comment = False)

**Review by Jerry-GeGarcia — APPROVED** [review]
:shipit:

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (nihemstr) [inline src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:121]: Do we use this in counting or binary? Does this protect the actual state variable? We should acquire the semaphore at the beginning and have a way to block any operations after assuming this is meant for shared memory.

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (nihemstr) [inline src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:125]: I don't think we need to log when this is called. I'm not sure how we'd use this to determine any issue given we're either uninitialized (which might be a good debug trace at the beginning of the function) or we can see the trace where we've already uninited the state before and we're suddenly trying again?

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (nihemstr) [inline src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:129]: We should do a copy of the value here and not take a constant pointer to something that can be changed in other sections of the code outside of the semaphore.

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (nihemstr) [inline src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:191]: All of these getters should return copies of the value instead of the actual value or the callers need to lock the semaphore and manage access themselves.

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (nihemstr) [inline src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:246]: maybe nit: do we know these will always be wellformed strings? Do we need to do ADUC_StrCopyN()

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (nihemstr) [inline src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:299]: is there a call to this function that would never have subscribed set to true?

File: src/extensions/agent_modules/shared/state_store/tests/agent_state_store_ut.cpp
Comment (nihemstr) [inline src/extensions/agent_modules/shared/state_store/tests/agent_state_store_ut.cpp:23]: May not be apart of this PR but we should start being more aggressive about happy + unhappy paths making sure we're exiting the way we want.

File: src/utils/c_utils/src/string_c_utils.c
Comment (nihemstr) [inline src/utils/c_utils/src/string_c_utils.c:456]: Add IsNullOrEmpty(src)

**Review by nihemstr — CHANGES_REQUESTED** [review]
Added some comments. Not really sure how to approach reading the overall flow but was able to get through some of the bigger stuff in this PR.

The ADUC_Store getters and setters running around the semaphores needs to be changed for the operations to prevent misreads / miswrites (eg someone is setting the deviceId while another person is reading the const char* value returned by the function).

File: src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c:840]: I changed these to messageId in the latest PR highest in the stack

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/inc/aduc/adu_agent_info_management.h
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/inc/aduc/adu_agent_info_management.h:22]: I'll do a separate PR to touch and clang-format everything before merging into feature branch. I didn't want to generate lots of clang-format noise in these stacked ones because all the new files basically had needed it.

**Comment by jw-msft** [comment]
> > * For 'ainfo_resp' messages, if the Correlation Data matches the 'en,the client should parse the message and update the agentinfo state.
> 
> what is 'en? Oh, enrollment.
> 
> Refers to: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c:65 in [4720e51](https://github.com/Azure/iot-hub-device-update/commit/4720e512a20cf4eb0d01d8f7f504bbe837449c27). [](commit_id = [4720e51](https://github.com/Azure/iot-hub-device-update/commit/4720e512a20cf4eb0d01d8f7f504bbe837449c27), deletion_comment = False)

No actually it's a serious typo--I think I was trying to say `matches, then the client` but I was typing way to fast with probably some editor key-seq macros mixed in.

File: src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c:817]: It is currently thread-safe as there's only one thread as it stems from mosquitto_loop() call after DoWork call from main() that would result in callbacks like OnConnect, OnPublish, OnMessage, etc.  I am adding another thread to handle items on the workqueue in an upcoming PR but the additional thread will update state store with update result reporting json and then the main thread will pick that up and interact with libmosquitto to send the updrslt_req, handle puback and adu service updrslt_resp response.

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:299]: Good point. Not currently because in ADUC_Communication_Channel_OnSubscribe() it ignores the reason code--if it's a failure reason code, then it should call this with false so that it's not subscribed because the SUBACK had a failure.

Opened this bug on it to track fixing it:

[Bug 48162686](https://microsoft.visualstudio.com/OS/_workitems/edit/48162686): [Gen2] ADUC_Communication_Channel_OnSubscribe needs to check mqtt reason code of SUBACK to know when topic subscribe failed

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c]: yes. Added with this commit: [add tracing for puback failure and clang-format](https://github.com/Azure/iot-hub-device-update/pull/584/commits/ba38bbca78bf548ccec83aa3ef9cd0cc93b6c9e1)

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_enrollment_management.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_enrollment_management.c:129]: Factored out into adu_mqtt_common.c now with this commit: [factor out puback handling to adu_mqtt_common](https://github.com/Azure/iot-hub-device-update/pull/584/commits/bd1513cbd9eadca3c66734cad1cfdf88fe227371)

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/agentinfo_request_operation.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/agentinfo_request_operation.c:222]: It is currently a default setting in the client side, but we are having conversations about also having poll freq and retry/timeout settings flow down from service.

File: src/extensions/agent_modules/shared/common/inc/aduc/adu_module_state.h
Comment (jw-msft) [reply src/extensions/agent_modules/shared/common/inc/aduc/adu_module_state.h]: Fixed with this commit: [fix comment - we only](https://github.com/Azure/iot-hub-device-update/pull/584/commits/a58ea8c43d782d7df92582bb216e752b26326c8f)

File: src/extensions/agent_modules/shared/common/inc/aduc/adu_module_state.h
Comment (jw-msft) [reply src/extensions/agent_modules/shared/common/inc/aduc/adu_module_state.h]: went with imperative (we only subscribe)

File: src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c]: done with commit: [move LOG_ALL_MOSQUITTO define to top](https://github.com/Azure/iot-hub-device-update/pull/584/commits/e52c96e7eaaecc06290575209cf4809ea9e9f3a1)

File: src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c]: Resolved with commit: [move set of mosqResult success right before done label](https://github.com/Azure/iot-hub-device-update/pull/584/commits/dcbb183f522fb604b3083a20b681ba726d266eef)

File: src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/communication_channel/mqtt_broker/src/adu_communication_channel.c:1376]: as discussed offline, VECTOR_push_back copies the item's struct memory so effectively is a shallow copy of fields and, thus, transfers ownership of ptr fields.

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/state_store/src/agent_state_store.c]: good catch. the pointers need an alloc and deep copy in general. Fixed with this commit:[make copies of strings for state store getters](https://github.com/Azure/iot-hub-device-update/pull/584/commits/8635ed011ebfc7396ecc82ffe15532439285e6df)

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/state_store/src/agent_state_store.c]: Good catch. Now doing a ADUC_StrNLen check and returning ADUC_STATE_STORE_RESULT_ERROR if == 0 or == MAX_ADU_SCOPE_ID_BYTE_LEN. Also, copying with MallocAndSubstr for now since it has N bound and not as churned as ADUC_StrCopyN with later PRs (less conflicts for now).

Here is the commit that addresses this: [bound state store set of scopeId by max scopeId bytelen](https://github.com/Azure/iot-hub-device-update/pull/584/commits/720d02ee1e2ef727b0e93efa3cd56824667303f8)

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c]: Fixed with following commit: [use size_t in fn sig in lieu of uint16_t](https://github.com/Azure/iot-hub-device-update/pull/584/commits/fce31ee12f3c577aa9df76feac82638636ac45a1)

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/state_store/src/agent_state_store.c]: addressed with commit: [make copies of strings for state store getters](https://github.com/Azure/iot-hub-device-update/pull/584/commits/8635ed011ebfc7396ecc82ffe15532439285e6df)

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/state_store/src/agent_state_store.c]: removed with commit: [omit trace when called and already deinit'ed](https://github.com/Azure/iot-hub-device-update/pull/584/commits/5bf0362a8871f2cd51c3aae2e86c9fc4ff8fa46c)

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:128]: By design. No need to acquire the semaphore becaue this will only be called after we exit the loop that was calling the doWork on the agent modules so nothing and also having stopped the worker threads, so it would not be competing with any other thread. For the same reason, no need to block operations after. It should probably just be a mutex and will change it to that later.

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:82]: Yes, I believe that is removed in a later PR in the stack.

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:68]: I plan on removing this if haven't already in later PR in the stack.

File: src/extensions/agent_modules/shared/state_store/src/agent_state_store.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/state_store/src/agent_state_store.c:26]: The technical max number of utf-8 bytes for topics (according to MQTT spec) is 65536 bytes, so I've increased it to that so that we do not artificially preclude future scenarios.

I've also updated `MAX_ADU_DEVICE_ID_BYTE_LEN` and `MAX_ADU_SCOPE_ID_BYTE_LEN` based on tracking those down.

File: src/extensions/agent_modules/shared/mosquitto_utils/src/adu_mosquitto_utils.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/mosquitto_utils/src/adu_mosquitto_utils.c]: Thanks. Fixed with commit: [cast uint16_t to size_t for ADUC_AllocAndStrCopyN](https://github.com/Azure/iot-hub-device-update/pull/584/commits/a70cf7d0ef19ae5d40e0f80d469899f339af411c)

File: src/extensions/agent_modules/shared/mosquitto_utils/src/adu_mosquitto_utils.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/mosquitto_utils/src/adu_mosquitto_utils.c:96]: not magic, just mosquitto_property_read_binary requires void** for this argument and the compiler won't allow passing a char** to it and the char is 8 bits so perfect for bytes of binary data. Under the hood it uses calloc and memcpy that both return and take void* so I suppose the API author didn't want to mess with the void* typing by casting things to char*.

I've summarized all this in a comment in this commit: [add comment about void** passed to mosquitto_property_read_property](https://github.com/Azure/iot-hub-device-update/pull/584/commits/8deef67a05e0751d1e7e81ca476e1f8ad067ab94)

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c:118]: done with this commit: [prefix ParseAndValidateCommon.. with ADU_MosquittoUtils_](https://github.com/Azure/iot-hub-device-update/pull/584/commits/bb3b2c9e383424e229ecfa1d32fdb45d1c4a7fb7)

Opened this tracking bug to prefix the others: [Bug 48195268](https://microsoft.visualstudio.com/OS/_workitems/edit/48195268): [Gen2] Make Consistent ADU_MosquittoUtils_ Prefix for all functions in adu_mosquitto_utils

File: src/extensions/agent_modules/shared/mosquitto_utils/src/adu_mosquitto_utils.c
Comment (jw-msft) [reply src/extensions/agent_modules/shared/mosquitto_utils/src/adu_mosquitto_utils.c]: I've made any arguments (in or out) size_t but for locals that need to interact with mosquitto APIs I've kept them as-is.  Let me know if you feel strongly about it and I can revisit.

File: src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c
Comment (jw-msft) [reply src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c]: Fixed with commit: [mv corr id test after fn arg validation](https://github.com/Azure/iot-hub-device-update/pull/584/commits/9786d55efcea2b187b79611712c2f6ed92527163)

File: src/utils/c_utils/src/string_c_utils.c
Comment (jw-msft) [reply src/utils/c_utils/src/string_c_utils.c:461]: won't fix. There is unit test for empty src and it's safe due to use of ADUC_Safe_StrCopyN below and it will fail if srcLen was > 0. Anyways, these functions are revamped a bit in a later stack PR for branch that was merged into user/jw-msft/v2-procupd

**Comment by jw-msft** [comment]
merged up to top pr stack that was merged to topic

**Pull request closed without merging**
